### PR TITLE
Feature/waypoints urls

### LINKED ISSUE #516: Put waypoints in URL
Passed to OTP as `intermediatePlaces` parameter, repeated for each waypoint. Handle interactions with local storage and shared directions.

### PR DESCRIPTION
**Pull request by flibbertigibbet**
Retain waypoints in URL and handle parsing them. Closes #516.

To test, create a trip with waypoints, click to share link, and open the shortened shareable URL in another browser or incognito tab (that does not have site settings in local storage). Trip directions should load with waypoints showing, and from/to text correctly displayed.

### CODE REVIEW

File: src/app/scripts/cac/pages/cac-pages-directions.js [hunk 1]
```diff
@@ -23,17 +23,55 @@ CAC.Pages.Directions = (function ($, _, DirectionsList, Itinerary, Settings, Use
...
+
+        // Note: date/time does not get passed and so always defaults to departing now
+
+        var rawParams = location.search.slice(1).split('&');
+        var itineraryIndex = -1; // initialize to flag value for missing index parameter
+
+        // process parameters for consumption by OpenTripPlanner:
+        // strip out itineraryIndex, and convert singe waypoint parameter into
+        // multiple intermediatePlaces parameters.
+        var otpParams = _.map(rawParams, function(paramStr) {
```
Comment (flibbertigibbet) [inline src/app/scripts/cac/pages/cac-pages-directions.js]: I don't feel super about all the parameter-munging happening here; open to any suggestions on how to make this cleaner.

File: src/app/scripts/cac/pages/cac-pages-directions.js [hunk 2]
```diff
@@ -48,11 +86,13 @@ CAC.Pages.Directions = (function ($, _, DirectionsList, Itinerary, Settings, Use
...
-            data: params
+            data: otpParams,
+            processData: false
         }).then(function(data) {
             var itineraries = data.plan.itineraries;
-            var params = data.requestParameters;
-            var itinerary = new Itinerary(itineraries[itineraryIndex], itineraryIndex, params);
+            var itinerary = new Itinerary(itineraries[itineraryIndex],
+                                          itineraryIndex,
+                                          otpParams);
```
Comment (KlaasH) [inline src/app/scripts/cac/pages/cac-pages-directions.js]: 93184634e gets rid of the third argument to `Itinerary`.

File: src/app/scripts/cac/routing/cac-routing-plans.js [hunk 3]
```diff
@@ -49,7 +49,7 @@ CAC.Routing.Plans = (function($, moment, _, UserPreferences, Itinerary, Settings
 
                 // return the Itinerary objects for the unique collection
                 var itineraries = _(planItineraries).map(function(itinerary, i) {
-                    return new Itinerary(itinerary, i, data.requestParameters);
+                    return new Itinerary(itinerary, i, urlParams);
```
Comment (KlaasH) [inline src/app/scripts/cac/routing/cac-routing-plans.js]: As above, `Itinerary` doesn't take the third argument anymore.

File: src/app/scripts/cac/pages/cac-pages-directions.js (hunk 1 above)
Comment (KlaasH) [inline src/app/scripts/cac/pages/cac-pages-directions.js]: I also do not love all the `indexOf` footwork.  Here's a version that breaks up the initial param string further, into an array of `[paramName, value]`arrays, then defines rewriting rules and parses the array using them.

It's not necessarily shorter, and it's not devastatingly readable, but I think it's more direct and more flexible.

```
var rawParams = _.map(location.search.slice(1).split('&'), function(p) { 
    return p.split('=');
});

// Pull out the itinerary index and validate it
var itineraryIndex = _.remove(rawParams, function (p) { return p[0] === 'itineraryIndex';});
if (itineraryIndex.length !== 1) {
    console.error('Must specify itineraryIndex URL parameter');
    return;
} else {
    itineraryIndex = parseInt(itineraryIndex[0][1]);
    if (isNaN(itineraryIndex)) {
        console.error('itineraryIndex URL parameter must be an integer');
        return;
    }
}

// Rules for rewriting GoPhillyGo URL params to OTP query params.
// If 'values' is defined, parse out the value into multiple copies of the param,
// one per item in the array returned by calling 'values' with the URL parameter value.
// Any URL param not in the mapping is passed through unchanged.
var mapping = {
    origin: { paramName: 'fromPlace' },
    originText: { paramName: 'fromText' },
    destination: { paramName: 'toPlace' },
    destinationText: { paramName: 'toText' },
    waypoints: {
        paramName: 'intermediatePlaces',
        values: function (val) { 
            return _.map(decodeURIComponent(val).split(';'), encodeURIComponent);
        }
    },
};

// Rewrites an array of [URLparam, value] pairs into an array of [OTPparam, value] pairs,
// using the mapping above.
var otpParams = _(rawParams).map(function (param) {
    if (!mapping[param[0]]) {
        return [param];
    } else {
        var config = mapping[param[0]];
        if (!config.values) {
            return [[config.paramName, param[1]]];
        } else {
            return _.map(config.values(param[1]), function (val) {
                return [config.paramName, val];
            });
        }
    }
}).flatten().value();
// Join array into param string
otpParams = _(otpParams).map(function (item) { return item.join('='); }).join('&');
```

**Pull request merged**